Mathlib/RingTheory/Ideal/IsPrincipal.lean

```diff
@@ -0,0 +1,89 @@
+/-
+Copyright (c) 2024 Xavier Roblot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Xavier Roblot
+-/
+import Mathlib.RingTheory.Ideal.Operations
+
+/-!
+# Principal Ideals
+
+This file deals with the set of principal ideals of a `CommRing R`.
+
+## Main definitions and results
+
+* `Ideal.isPrincipalSubmonoid`: the submonoid of `Ideal R` formed by the principal ideals of `R`.
+
+* `Ideal.associatesMulEquivIsPrincipal`: the `MulEquiv` between the monoid of `Associates R` and
+the submonoid of principal ideals of `R`.
+
+-/
+
+variable {R : Type*} [CommRing R]
+
+namespace Ideal
+
+open Submodule
+
+variable (R) in
+/-- The principal ideals of `R` form a submonoid of `Ideal R`. -/
+def isPrincipalSubmonoid : Submonoid (Ideal R) where
+  carrier := { I | IsPrincipal I}
+  mul_mem' := by
+    rintro _ _ ⟨x, rfl⟩ ⟨y, rfl⟩
+    exact ⟨x * y, Ideal.span_singleton_mul_span_singleton x y⟩
+  one_mem' := ⟨1, one_eq_span⟩
+
+theorem mem_isPrincipalSubmonoid_iff {I : Ideal R} :
+    I ∈ isPrincipalSubmonoid R ↔ IsPrincipal I := Iff.rfl
+
+theorem span_singleton_mem_isPrincipalSubmonoid (a : R) :
+    span {a} ∈ isPrincipalSubmonoid R := mem_isPrincipalSubmonoid_iff.mpr ⟨a, rfl⟩
+
+variable [IsDomain R]
+
+variable (R) in
+/-- The equivalence between `Associates R` and the principal ideals of `R` defined by sending the
+class of `x` to the principal ideal generated by `x`. -/
+noncomputable def associatesEquivIsPrincipal :
+    Associates R ≃ {I : Ideal R // IsPrincipal I} := by
+  refine Equiv.ofBijective (fun x ↦ ⟨span {Quot.out x}, ⟨Quot.out x, rfl⟩⟩)
```

Comment:
Can the inverse be given by `Submodule.IsPrincipal.generator`?

Reply:
Yes. I didn't do it in the first place since it requires to import `Mathlib.RingTheory.PrincipalIdealDomain` but now that this is in its own file, I guess it's okay. I added also `associatesEquivIsPrincipal_symm_apply`. 

Thanks for the suggestion!

Reply:
I meant something like
```
variable (R) in
/-- The equivalence between `Associates R` and the principal ideals of `R` defined by sending the
class of `x` to the principal ideal generated by `x`. -/
noncomputable def associatesEquivIsPrincipal :
    Associates R ≃ {I : Ideal R // IsPrincipal I} where
  toFun := Quotient.lift (fun x ↦ ⟨span {x}, x, rfl⟩)
    (fun a b e ↦ by simpa [span_singleton_eq_span_singleton])
  invFun I := .mk I.2.generator
  left_inv := Quotient.ind fun r ↦ by simpa using
    Ideal.span_singleton_eq_span_singleton.mp (@Ideal.span_singleton_generator _ _ _ ⟨_, rfl⟩)
  right_inv I := by simp
```

Reply:
This makes most of the lemmas below `rfl`.

Reply:
Oh, right. I went a bit overboard here.